java/ql/src/experimental/Security/CWE/CWE-522/InsecureBasicAuth.ql

```diff
@@ -0,0 +1,264 @@
+/**
+ * @name Insecure basic authentication
+ * @description Basic authentication only obfuscates username/password in Base64 encoding, which can be easily recognized and reversed. Transmission of sensitive information not over HTTPS is vulnerable to packet sniffing.
+ * @kind path-problem
+ * @id java/insecure-basic-auth
+ * @tags security
+ *       external/cwe-522
+ *       external/cwe-319
+ */
+
+import java
+import semmle.code.java.frameworks.Networking
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * The Java class `org.apache.http.client.methods.HttpRequestBase`. Popular subclasses include `HttpGet`, `HttpPost`, and `HttpPut`.
+ * And the Java class `org.apache.http.message.BasicHttpRequest`.
+ */
+class ApacheHttpRequest extends RefType {
+  ApacheHttpRequest() {
+    this
+        .getASourceSupertype*()
+        .hasQualifiedName("org.apache.http.client.methods", "HttpRequestBase") or
+    this.getASourceSupertype*().hasQualifiedName("org.apache.http.message", "BasicHttpRequest")
+  }
+}
+
+/**
+ * Class of Java URL constructor.
+ */
+class URLConstructor extends ClassInstanceExpr {
+  URLConstructor() { this.getConstructor().getDeclaringType() instanceof TypeUrl }
+
+  // URLs constructed with the string constructor `URL(String spec)`
+  Expr specArg() {
+    this.getConstructor().getParameter(0).getType() instanceof TypeString and
+    (
+      this.getConstructor().getNumberOfParameters() = 1 and
+      result = this.getArgument(0) // First argument contains the whole spec.
+    )
+  }
+
+  // URLs constructed with any of the three string constructors below:
+  // `URL(String protocol, String host, int port, String file)`,
+  // `URL(String protocol, String host, int port, String file, URLStreamHandler handler)`,
+  // `URL(String protocol, String host, String file)`
+  Expr protocolArg() {
+    this.getConstructor().getParameter(0).getType() instanceof TypeString and
+    (
+      this.getConstructor().getNumberOfParameters() > 1 and
+      result = this.getArgument(0) // First argument contains the protocol part.
+    )
+  }
+
+  Expr hostArg() {
+    this.getConstructor().getParameter(0).getType() instanceof TypeString and
+    (
+      this.getConstructor().getNumberOfParameters() > 1 and
+      result = this.getArgument(1) // Second argument contains the host part.
+    )
+  }
+}
+
+/**
+ * Gets a regular expression for matching private hosts.
+ */
+private string getPrivateHostRegex() {
+  result = "localhost(/.*)?" or
```

Comment:
You could make a single regex by joining them with `|`

Reply:
I've changed to a single regex as per the suggestion while keeping each domain pattern separate.